### DISCUSSION TITLE
Implementing predicted  FvSource terms in OpenFOAM/DAFoam

### DISCUSSION THREAD

**Discussion by aursetto-ops** [mdolab/dafoam#948] [General]
Hi all,

I have developed a data-driven workflow using DAFoam for Data Assimilation (DA) on LES and RANS data. I extracted the fvSource and velocity gradients (∇U) to train a Neural Network (NN) based on Pope Tensors.

Now, I am attempting to run a RANS simulation by injecting the fvSource values as a corrective source term in the momentum equation. However, using fvOptions (e.g., explicitSource) does not seem to affect the solution.
How can I ensure the solver reads fvSource from the 0/ directory and applies it ?

**Comment by friedenhe** [comment]
The current implementation does not read fvSources from a file. Check this [line](https://github.com/mdolab/dafoam/blob/7fb01fe4cddc972ca51b6c083ba91b7261cf5c92/src/adjoint/DASolver/DASimpleFoam/createFieldsSimple.H#L50). You can change `NO_READ` to `READ_IF_PRESENT` and recompile DAFoam. If it works, can you submit a PR to the DAFoam repo for this new feature?

**Reply by aursetto-ops** [reply]
Thank you for your response.

I've been testing a different approach: I placed the fvSource field (extracted from my data-assimilation) in the 0/ directory of a new case and added an fvOptions file to handle it. I’m just running the case with simpleFoam directly, without using any runscripts.

Here is how I defined my fvOptions:


FoamFile
{
    version     2.0;
    format      ascii;
    class       dictionary;
    location    "system";
    object      fvOptions;
}
// * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * //
momentumSource
{
    type            vectorCodedSource;
    active          yes;

    name            myFvSource;     

    vectorCodedSourceCoeffs
    {
        selectionMode   all;
        fields          (U);

        codeInclude
        #{
            #include "fvCFD.H"
        #};

        codeCorrect
        #{
         
        #};

        codeAddSup
        #{
            if (!mesh_.foundObject<volVectorField>("fvSource"))
                {
                    volVectorField* fvSourcePtr = new volVectorField
                    (
                        IOobject
                        (
                            "fvSource",
                            "0",
                            mesh_,
                            IOobject::MUST_READ,
                            IOobject::NO_WRITE
                        ),
                        mesh_
                    );
                    mesh_.objectRegistry::store(fvSourcePtr);
                }

                const volVectorField& fvSrc = mesh_.lookupObject<volVectorField>("fvSource");

                static bool dbgPrinted = false;
                if (!dbgPrinted && Pstream::master())
                {
                    const label nLocal = fvSrc.internalField().size();
                    const label nPrint = (nLocal < 5) ? nLocal : 5;

                    Info<< "\n[DEBUG] fvSource instance = " << fvSrc.instance()
                        << ", path = " << fvSrc.objectPath() << nl
                        << "[DEBUG] nLocalCells = " << nLocal << nl;

                    for (label i = 0; i < nPrint; ++i)
                    {
                        Info<< "[DEBUG] fvSource[" << i << "] = " << fvSrc[i] << nl;
                    }

                    dbgPrinted = true;
                }
                vectorField& source = eqn.source();
                vector totAdded = vector::zero;
                
                forAll(cells_, i)
                {
                    const label cellI = cells_[i];
                    const vector dS = fvSrc[cellI]*mesh_.V()[cellI];  
                    source[cellI] -= dS;
                    totAdded += dS;
                }
                
                reduce(totAdded, sumOp<vector>());
                
                if (Pstream::master())
                {
                    Info<< "[DEBUG] Integral( fvSource * V ) = " << totAdded << nl;
                }

        #};

        codeSetValue
        #{
        #};
    }
}
